ml-algorithms/src/main/java/org/opensearch/ml/engine/ingest/S3DataIngestion.java

```diff
@@ -0,0 +1,207 @@
+        for (int sourceIndex = 0; sourceIndex < s3Uris.size(); sourceIndex++) {
+            successRates.add(ingestSingleSource(s3, s3Uris.get(sourceIndex), mlBatchIngestionInput, sourceIndex, isSoleSource));
+        }
+
+        return calculateSuccessRate(successRates);
+    }
+
+    public double ingestSingleSource(
+        S3Client s3,
+        String s3Uri,
+        MLBatchIngestionInput mlBatchIngestionInput,
+        int sourceIndex,
+        boolean isSoleSource
+    ) {
+        String bucketName = getS3BucketName(s3Uri);
+        String keyName = getS3KeyName(s3Uri);
+        GetObjectRequest getObjectRequest = GetObjectRequest.builder().bucket(bucketName).key(keyName).build();
+        double successRate = 0;
+
+        try {
+            ResponseInputStream<GetObjectResponse> s3is = AccessController
+                .doPrivileged((PrivilegedExceptionAction<ResponseInputStream<GetObjectResponse>>) () -> s3.getObject(getObjectRequest));
+            BufferedReader reader = new BufferedReader(new InputStreamReader(s3is, StandardCharsets.UTF_8));
+
+            List<String> linesBuffer = new ArrayList<>();
+            String line;
+            int lineCount = 0;
+            // Atomic counters for tracking success and failure
+            AtomicInteger successfulBatches = new AtomicInteger(0);
+            AtomicInteger failedBatches = new AtomicInteger(0);
+            // List of CompletableFutures to track batch ingestion operations
+            List<CompletableFuture<Void>> futures = new ArrayList<>();
+
+            while ((line = reader.readLine()) != null) {
+                linesBuffer.add(line);
+                lineCount++;
+
+                // Process every 100 lines
+                if (lineCount == 100) {
+                    // Create a CompletableFuture that will be completed by the bulkResponseListener
+                    CompletableFuture<Void> future = new CompletableFuture<>();
+                    batchIngest(
+                        linesBuffer,
+                        mlBatchIngestionInput,
+                        getBulkResponseListener(successfulBatches, failedBatches, future),
+                        sourceIndex,
+                        isSoleSource
+                    );
+
+                    futures.add(future);
+                    linesBuffer.clear();
+                    lineCount = 0;
+                }
+            }
+            // Process any remaining lines in the buffer
+            if (!linesBuffer.isEmpty()) {
+                CompletableFuture<Void> future = new CompletableFuture<>();
+                batchIngest(
+                    linesBuffer,
+                    mlBatchIngestionInput,
+                    getBulkResponseListener(successfulBatches, failedBatches, future),
+                    sourceIndex,
+                    isSoleSource
+                );
+                futures.add(future);
+            }
+
+            reader.close();
```

Comment:
If any exception happes before this line, reader will never be closed. Use try (reader = ...) ?

Reply:
To be on the safe side I just added it into the try(). My 2 cents is that this shouldn't be a problem even it exceptions out of this code early.